alacritty/src/cursor.rs

```diff
@@ -1,112 +1,66 @@
-    cursor_thickness: f64,
-) -> RasterizedGlyph {
-    // Calculate the cell metrics.
-    //
-    // NOTE: With Rust 1.47+ `f64 as usize` is defined to clamp automatically:
-    // https://github.com/rust-lang/rust/commit/14d608f1d8a0b84da5f3bccecb3efb3d35f980dc
-    let height = (metrics.line_height + f64::from(offset_y)).max(1.) as usize;
-    let mut width = (metrics.average_advance + f64::from(offset_x)).max(1.) as usize;
-    let line_width = (cursor_thickness * width as f64).round().max(1.) as usize;
-
-    // Double the cursor width if it's above a double-width glyph.
-    if is_wide {
-        width *= 2;
-    }
+    pub color: Rgb,
 
-    match cursor {
-        CursorShape::HollowBlock => get_box_cursor_glyph(height, width, line_width),
-        CursorShape::Underline => get_underline_cursor_glyph(width, line_width),
-        CursorShape::Beam => get_beam_cursor_glyph(height, line_width),
-        CursorShape::Block => get_block_cursor_glyph(height, width),
-        CursorShape::Hidden => RasterizedGlyph::default(),
-    }
-}
+    pub key: CursorKey,
 
-/// Return a custom underline cursor character.
-pub fn get_underline_cursor_glyph(width: usize, line_width: usize) -> RasterizedGlyph {
-    // Create a new rectangle, the height is relative to the font width.
-    let buf = vec![255u8; width * line_width * 3];
-
-    // Create a custom glyph with the rectangle data attached to it.
-    RasterizedGlyph {
-        c: ' ',
-        top: line_width as i32,
-        left: 0,
-        height: line_width as i32,
-        width: width as i32,
-        buf: BitmapBuffer::RGB(buf),
-    }
+    pub thickness: f64,
 }
 
-/// Return a custom beam cursor character.
-pub fn get_beam_cursor_glyph(height: usize, line_width: usize) -> RasterizedGlyph {
-    // Create a new rectangle that is at least one pixel wide
-    let buf = vec![255u8; line_width * height * 3];
-
-    // Create a custom glyph with the rectangle data attached to it
-    RasterizedGlyph {
-        c: ' ',
-        top: height as i32,
-        left: 0,
-        height: height as i32,
-        width: line_width as i32,
-        buf: BitmapBuffer::RGB(buf),
+impl Cursor {
+    pub fn new(x: f32, y: f32, color: Rgb, thickness: f64, key: CursorKey) -> Self {
+        Self { x, y, color, thickness, key }
     }
-}
 
-/// Returns a custom box cursor character.
-pub fn get_box_cursor_glyph(height: usize, width: usize, line_width: usize) -> RasterizedGlyph {
-    // Create a new box outline rectangle.
-    let mut buf = Vec::with_capacity(width * height * 3);
-    for y in 0..height {
-        for x in 0..width {
-            if y < line_width
-                || y >= height - line_width
-                || x < line_width
-                || x >= width - line_width
-            {
-                buf.append(&mut vec![255u8; 3]);
-            } else {
-                buf.append(&mut vec![0u8; 3]);
-            }
+    pub fn into_rects(self, size_info: &SizeInfo) -> Vec<RenderRect> {
```

Comment:
Multiple rects for the box cursor, correct?

Reply:
Yes, however what do you think about passing a mutable reference to a vector? I think it should be more clear, and especially since reduce allocations on each draw.

Reply:
I'm not sure what you mean, why would it need to be mutable?

And the Vecs are created in this function, right? So that would create lifetime problems. Or do you plan to store the vector on the `Cursor`?